Sources/Algorithms/LazySplitSequence.swift

```diff
@@ -0,0 +1,321 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// A sequence that lazily splits a base sequence into subsequences separated by
+/// elements that satisfy the given `whereSeparator` predicate.
+///
+/// - Note: This type is the result of
+///
+///     x.split(maxSplits:omittingEmptySubsequences:whereSeparator)
+///     x.split(separator:maxSplits:omittingEmptySubsequences)
+///
+///   where `x` conforms to `LazySequenceProtocol`.
+public struct LazySplitSequence<Base: Sequence> {
+  internal let base: Base
+  internal let isSeparator: (Base.Element) -> Bool
+  internal let maxSplits: Int
+  internal let omittingEmptySubsequences: Bool
+}
+
+extension LazySplitSequence {
+  public struct Iterator {
+    public typealias Element = [Base.Element]
+
+    internal var base: Base.Iterator
+    internal let isSeparator: (Base.Element) -> Bool
+    internal let maxSplits: Int
+    internal let omittingEmptySubsequences: Bool
+    internal var separatorCount = 0
+    internal var sequenceLength = 0
+
+    internal init(
+      base: Base.Iterator,
+      whereSeparator: @escaping (Base.Element) -> Bool,
+      maxSplits: Int,
+      omittingEmptySubsequences: Bool
+    ) {
+      self.base = base
+      self.isSeparator = whereSeparator
+      self.maxSplits = maxSplits
+      self.omittingEmptySubsequences = omittingEmptySubsequences
+    }
+  }
+}
+
+extension LazySplitSequence.Iterator: IteratorProtocol {
+  public mutating func next() -> Element? {
+    /// Separators mark the points where we want to split (cut in two) the base
+    /// collection, removing the separator in the process.
+    ///
+    /// Each split yields two subsequences, though splitting at the start or end
+    /// of a sequence yields an empty subsequence where there were no elements
+    /// adjacent to the cut.
+    ///
+    /// Thus the maximum number of subsequences returned after iterating the
+    /// entire base collection (including empty ones, if they are not omitted)
+    /// will be at most one more than the number of splits made (equivalently,
+    /// one more than the number of separators encountered).
+    ///
+    /// The number of splits is limited by `maxSplits`, and thus may be less
+    /// than the total number of separators in the base collection.
+    ///
+    ///     [1, 2, 42, 3, 4, 42, 5].split(separator: 42,
+    ///                                   omittingEmptySubsequences: false)
+    ///     // first split -> [1, 2], [3, 4, 42, 5]
+    ///     // last split  -> [1, 2], [3, 4], [5]
+    ///
+    ///     [1, 2, 42, 3, 4, 42, 5, 42].split(separator: 42,
+    ///                                       maxSplits: 2,
+    ///                                       omittingEmptySubsequences: false)
+    ///     // first split -> [1, 2], [3, 4, 42, 5, 42]
```

Comment:
Should the last `42` appear in the result here? If I understand correctly it should be an empty subsequence?

Reply:
That actually meant for the line bellow (line 79) last split =]

Reply:
In this case with `maxSplits` == 2, the base sequence is only split twice, even though it contains three separators. Although the docs for the existing eager `split`s say, "Elements that are used to split the sequence are not returned as part of any subsequence", that is not how those methods actually work. In the case where `maxSplits` is less than the number of separators, the remaining separators do appear in the last subsequence.

Reply:
I dunno, is that a bug? Is the documentation saying exactly how it should be working?

Reply:
Ah no sorry, I think it was just a confusion from my part I see the point now =]